php-frontend/src/main/java/org/sonar/php/filters/SuppressWarningFilter.java

```diff
@@ -90,18 +101,47 @@ public void visitToken(SyntaxToken token) {
 
   private void processSuppressedWarningsInComment(SyntaxToken token, String comment) {
     Matcher matcher = SUPPRESS_WARNING_COMMENT_PATTERN.matcher(comment);
-    while (matcher.find()) {
-      String arguments = matcher.group("arguments");
-      Arrays.stream(arguments.split(","))
-        .map(str -> stripDoubleQuotes(str.trim()))
-        .forEach(ruleName -> {
-          for (int line = token.line(); line <= token.endLine(); line++) {
-            suppressedWarnings.addSuppressedWarning(getFileUri(), line, ruleName);
-          }
-        });
+
+    if (matcher.find()) {
+      PHPTree parent = findFarthestPhpTreeParent(token);
+      int startLine = Optional.ofNullable(parent)
+        .map(PHPTree::getFirstToken)
+        .map(SyntaxToken::line)
+        .orElse(token.line());
+      int endLine = Optional.ofNullable(parent)
+        .map(PHPTree::getLastToken)
+        .map(SyntaxToken::endLine)
+        .orElse(token.endLine());
+
+      do {
+        String arguments = matcher.group("arguments");
+        Arrays.stream(arguments.split(","))
+          .map(str -> stripDoubleQuotes(str.trim()))
+          .forEach(ruleName -> suppressedWarnings.addSuppressedWarning(getFileUri(), ruleName, startLine, endLine));
+      } while(matcher.find());
```

Comment:
Yes I'm not a big fan of `do{}while` also.
I needed it here because this code will be run for every comment, so to not decrease the performance I really don't want to compute the `PHPTree` and the range before it is really required.

Solution 2 doesn't seems good to me as it will require to compute those elements for each entry, or to move the entire logic into this static class, which is out of the scope of what it was designed for.

For solution 1, I gave it a second try, it's a bit better than the first time I wrote it thanks to the use of Range:
```
    PHPTree parent = null;
    Range<Integer> range = null;
    while (matcher.find()) {
      if (parent == null) {
        parent = findFarthestPhpTreeParent(token);
        range = computeLineRange(parent, token, token);
      }
      String arguments = matcher.group("arguments");
      Arrays.stream(arguments.split(","))
        .map(str -> stripDoubleQuotes(str.trim()))
        .forEach(ruleName -> suppressedWarnings.addSuppressedWarning(getFileUri(), ruleName, range));
    }
```
Doesn't seems to bad to me, we easily understand the `while` logic and the fact that we init some element on first loop, I will keep it that way 👍

Reply:
Ah, didn't work, `range` needs to be final or effectively final to be used in lambda expression :sweat_smile:
Will see If I can avoid using lambda expression here while keeping good clarity.